src/backends/torch/torchgraphbackend.cc

```diff
@@ -0,0 +1,242 @@
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * deepdetect is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with deepdetect.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+
+#include "torchgraphbackend.h"
+
+
+namespace dd
+{
+  using torch::nn::AnyModule;
+  using torch::nn::RNN;
+  using torch::nn::RNNOptions;
+  using torch::nn::LSTM;
+  using torch::nn::LSTMOptions;
+  using torch::nn::Linear;
+  using torch::nn::LinearOptions;
+  using torch::Tensor;
+
+
+  void TorchGraphBackend::set_input(torch::Tensor input)
+  {
+	std::vector<int> dim(input.sizes().begin(),input.sizes().end());
+	this->set_input_dim(dim);
+	_variables[_inputname] = input;
+	BaseGraph::finalize();
+	allocate_modules();
+  }
+
+  void TorchGraphBackend::finalize()
+  {
+	BaseGraph::finalize();
+	allocate_modules(false);
+  }
+
+
+  void TorchGraphBackend::finalize(at::IntArrayRef dim)
+  {
+	finalize(dim.vec());
+  }
+
+  void TorchGraphBackend::finalize(std::vector<int> dim)
+  {
+	this->set_input_dim(dim);
+	BaseGraph::finalize();
+	allocate_modules();
+  }
+
+
+  void TorchGraphBackend::finalize(std::vector<int64_t> dim)
+  {
+	std::vector<int> dimint;
+	for (auto d:dim)
+	  dimint.push_back(d);
+	finalize(dimint);
+  }
+
+
+
+
+  torch::Tensor TorchGraphBackend::forward(torch::Tensor inputTensor)
+  {
+	set_input(inputTensor);
+	bool output_computed = false;
+	for (BaseGraph::Vertex o: _sortedOps)
+	  {
+		std::vector<torch::Tensor> out = forward(o);
+		std::vector<BaseGraph::Vertex> outputVars = this->outputs(o);
+		for (unsigned int i=0; i<outputVars.size(); ++i)
+		  {
+			if (varname(outputVars[i]) == _outputname)
+			  output_computed = true;
+			_variables[varname(outputVars[i])] = out[i];
+		  }
+	  }
+	if (!output_computed)
+	  throw TorchGraphException("did not compute output, please check NN graph");
+	return _variables[_outputname];
+  }
+
+  std::vector<torch::Tensor> TorchGraphBackend::forward(BaseGraph::Vertex v)
+  {
+
+	std::vector<torch::Tensor> inputsTensor;
+	for (BaseGraph::Vertex vi: this->inputs(v))
+	  {
+		inputsTensor.push_back(_variables[varname(vi)]);
+	  }
+
+	std::vector<torch::Tensor> output;
+	std::string optype = this->optype(v);
+	if (optype == "LSTM" || optype == "RNN")
+	  {
+		//get<0>(f()) for all outputs / hidden
+		//get<0>(get<1>(f)) for last output
+		//get<1>(get<1>(f)) for last internal state
+		std::tuple<torch::Tensor, std::tuple<torch::Tensor, torch::Tensor>> full_output;
+		if (_lstm_continuation && _rnn_has_memories[opname(v)])
+		  {
+			full_output = _modules[opname(v)].
```

Comment:
Store `opname(v)` in variable to avoid calling it multiple times

Reply:
ok